src/inmanta/agent/agent.py

```diff
@@ -137,52 +139,136 @@ def __init__(self, scheduler: "ResourceScheduler", gid: uuid.UUID, reason: str)
         super().__init__(scheduler, dummy_id, gid, reason)
 
 
-class ResourceAction(ResourceActionBase):
-    def __init__(self, scheduler: "ResourceScheduler", resource: Resource, gid: uuid.UUID, reason: str) -> None:
+@dataclass(frozen=True)
+class ResourceInstallSpec:
+    """
+    This class encapsulates the requirements for a specific resource type for a specific model version.
+
+    :ivar resource_type: fully qualified name for this resource type e.g. std::File
+    :ivar model_version: the version of the model to use
+    :ivar pip_config: the pip config to use during requirements installation
+    :ivar requirements: python packages that must be installed prior to executing the module sources
+    :ivar sources: list of ModuleSource containing the code for deployment of this resource
+
+    """
+
+    resource_type: str
+    model_version: int
+    pip_config: PipConfig
+    requirements: Sequence[str]
+    sources: Sequence["ModuleSource"]
+
+
+class ResourceReference(dict[str, object]):
```

Comment:
this has to be a proper object, not just a dict. I would have an attributes field, an ID field an a requires field